parcel/package.json

```diff
@@ -1,7 +1,12 @@
 {
   "source": "index.html",
   "devDependencies": {
+    "@aurelia/runtime-html": /* @if !dev */"latest"/* @endif *//* @if dev */"dev"/* @endif */,
+    /* @if app-with-router */
+    "@aurelia/router": /* @if !dev */"latest"/* @endif *//* @if dev */"dev"/* @endif */,
+    /* @endif */
     "@aurelia/parcel-transformer": /* @if !dev */"latest"/* @endif *//* @if dev */"dev"/* @endif */,
+    "@parcel/config-default": "^2.6.0",
```

Comment:
https://github.com/parcel-bundler/parcel/issues/7999

I tested yarn pnp loose mode. It didn't help.
I really dislike this patch for parcel's issue.

Instead of doing here which unnecesarily affect users don't use yarn pnp, we can conditionally do "yarn add the-three" in after.js like you did for vscode integration.

Reply:
Since you commented that this also fixes webpack [here](https://github.com/aurelia/new/issues/85#issuecomment-1260225520), should I move those 2 dependencies in `common/package.json`?

Reply:
Yarn add will update package.json. But this will only update package.json for users who do uses pnp.

You can test the same approach for webpack with pnp, I have not tested that.

Reply:
To be honest, I am quite confused in the whole parcel and yarn pnp github issues. I am not interested to dig deep . Sorry :-) I had enough experience with yarn and pnpm.

I used yarn and pnpm both for quite some time few years ago, and contributed to them both. But finally I moved back to npm because I really don't want to spend my time to understand the difference anymore. The little saving on install time didn't justify the troubles (both known and unknown) I faced.

Reply:
@3cp The problem is - default skeletons could be (and could have been) one of the main reasons for poor `aurelia` adaption. I'm using `aurelia` since 2015 and there were always problems. If 30-50% of combinations for scaffolding do not work, that's not good.

Somewhere during my investigation I read that rule-of-thumb should be - you don't explicitly reference a module in your code which is not explicitly in your `package.json` dependencies. Possibly even linter rule... This makes sense to avoid error and applications unexpectedly breaking after update of packages.

In general, if `npm` was good enough, there would not be `yarn` and `pnpm`. If I'm not wrong they are now both shipped with `node` installation. Handled via [Corepack](https://nodejs.org/dist/latest/docs/api/corepack.html). So people will eventually move from `yarn` v1 (current is 3.2.3) and they will immediately run into all of those issues, same way I did.

Additional note - even after we solve the issue of adding extra packages, Parcel cannot do tree-shaking because of dynamically added module `@aurelia/runtime-html`. Maybe this modification should be done earlier in the pipeline, so that tree-shaking works. I mentioned this already, and you can see warning if you add `--trace-warnings` flag to `parcel` command(s).

Regarding `yarn add` in `after.js`, it's clear. I will do the change.

Reply:
One of the main troubles of au skeletons is due to we support too many combinations. Some of them like to introduce breaking changes very often :-)

Also I believe skeletons from many other communities assume npm in use.

I was also trying to reduce au2's skeleton options deliberately (comparing to au1). :-)

Reply:
@3cp opinion needed.

> You can test the same approach for webpack with pnp, I have not tested that.

So, as we discussed, I've been testing if adding additional packages by executing `yarn add` in `after.js` would fix the issues with Webpack. It didn't.

But additional problem appeared. While creating new project from the skeleton, because we, correctly, have in `package.json`:
```json
"typescript": "^4.7.3"
```
The latest stable version was picked up: `4.8.4`. It turned out that this version breaks Yarn's Visual Studio Code extension:
https://github.com/yarnpkg/berry/pull/4903.
The fix for `yarn dls @yarnpkg/sdks vscode` will be released next week, but this prompted me to look what is the root cause and it's Microsoft's refusal (2 years of discussion) to integrate `pnp`  module resolution directly, or to enable plugin infrastructure (for this I think they are right). Here are the details:
https://github.com/microsoft/TypeScript/issues/28289

So, after thinking about this, I have to agree with you, `Plug'n'Play` and `TypeScript` is a tricky business at the moment and things can break or at least have hiccups in the future.

Perhaps it would be safer approach to force `yarn` to create `node_modules`. This can be achieved by adding `.yarnrc.yml` file in the root of the project with the following entry:
>nodeLinker: node-modules

This configuration switch is working with both Parcel and Webpack. It enforces `node_modules` to be created, but at least it won't leave user with a newly created project broken. I also think we should keep `yarn start` and `pnpm start` display messages.

@3cp What do you think about this?

Reply:
The npm/yarn/pnpm question only pops up in after.js, and it's behind the check of silent mode.

Since deps install is not a mandatory step for scaffolding a project, this is not in questions.js.

When user ran makes in silent mode, we don't bother to ask user for deps installation. It's totally up to the user what to do next. The silent mode is designed for automation script, where we do as little as required.

The yarn config should work. We can present some help text (question choice has description) to clarify yarn pnp doesn't work well with au2.

This kind of ts/vscode/yarn issue is exactly what I complained about skeleton maintenance:-)

Reply:
You don't have to read package.json about ^2.6.0, it means any latest major version 2 at least higher then 2.6.0. It will always install latest parcel v2, not v2.6.0.

Just run yarn add foo, it will put ^latest.foo.version in package.json.

Parcel is not going to v3 for anytime soon.

Reply:
> The npm/yarn/pnpm question only pops up in after.js, and it's behind the check of silent mode.

You can simply default to `npm` in silent mode. Moving it to `questions.js` would allow to have files:
`yarn/.yarnrc.js` or `pnpm/.npmrc`, instead of  everything in `common` folder.

It would also allow, for example, in `typescript/package.json`:

```json
{
  "devDependencies": {
    // @if typescript && yarn-pnp
    "@parcel/config-default": "^2.6.0"
    // @endif
   "parcel": "^2.6.0"
```
So, all versions in `json` instead of `after.js`:
```js
await run('yarn', ["add", "@parcel/config-default@^2.6.0", "--dev"]);
```
Currently you have to update version `^2.6.0` in two different (non-intuitive) places.

However, this 2nd example is theoretical, because once we add `.yarnrc.yml` with `nodeLinker: node-modules`, we can remove handling of extra dependencies in `after.js` for Yarn's Plug'n'Play mode.

Moving the package-manager question might have pros and cons, so I am not arguing it's better. Just trying to make it clear.

> The yarn config should work. We can present some help text (question choice has description) to clarify yarn pnp doesn't work well with au2.

I had in mind to something along these lines:
```shell
npm
yarn (classic)
pnpm
```

Reply:
I mean we cannot assume user want to do npm install in silent mode. There is usage of private npm registry, user might want to touch .npmrc before deps install.

When design tool, I aid to have less assumption, that enables flexible use cases.

Putting them in question.js also means that's in our offical support. We have to cover yarn and pnpm in CI setup. But we are definitely not keen to officially support future versions of yarn and pnpm.

Reply:
Wait a min.
With `nodeLinker: node-modules`, is not that means we don't need to explicit install `@parcel/config-default`?

Reply:
Yes, I confirmed `nodeLinker: node-modules` means it all worked out with explicitly adding of deps.

Reply:
Why not :)
Our pnpm setup also uses patch in npmrc config yoo.